sel/src/Sel/SecretKey/AuthenticatedEncryption.hs

```diff
@@ -0,0 +1,388 @@
+    pure $ SecretKey (Foreign.castForeignPtr @CChar @CUChar secretKeyForeignPtr)
+
+-- | Convert a 'SecretKey' to a strict, hexadecimal-encoded 'Text'.
+--
+-- ⚠️  Be prudent as to where you store it!
+--
+-- @since 0.0.1.0
+secretKeyToHexText :: SecretKey -> Text
+secretKeyToHexText = Base16.encodeBase16 . secretKeyToBinary
+
+-- | Convert a 'SecretKey' to a strict, hexadecimal-encoded 'StrictByteString'.
+--
+-- ⚠️  Be prudent as to where you store it!
+--
+-- @since 0.0.1.0
+secretKeyToHexByteString :: SecretKey -> StrictByteString
+secretKeyToHexByteString = Base16.encodeBase16' . secretKeyToBinary
+
+-- | Convert a 'SecretKey' to a strict binary 'StrictByteString'.
+--
+-- ⚠️  Be prudent as to where you store it!
+--
+-- @since 0.0.1.0
+secretKeyToBinary :: SecretKey -> StrictByteString
+secretKeyToBinary (SecretKey secretKeyForeignPtr) =
+  BS.fromForeignPtr0
+    (Foreign.castForeignPtr @CUChar @Word8 secretKeyForeignPtr)
+    (fromIntegral @CSize @Int cryptoSecretboxKeyBytes)
+
+-- | A random number that must only be used once per exchanged message.
+-- It does not have to be confidential.
+-- It is of size 'cryptoSecretboxNonceBytes'.
+--
+-- @since 0.0.1.0
+newtype Nonce = Nonce (ForeignPtr CUChar)
+
+-- |
+--
+-- @since 0.0.1.0
+instance Eq Nonce where
+  (Nonce hk1) == (Nonce hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrEq hk1 hk2 cryptoSecretboxKeyBytes
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord Nonce where
+  compare (Nonce hk1) (Nonce hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrOrd hk1 hk2 cryptoSecretboxKeyBytes
+
+-- | ⚠️  Be prudent with what you do with it!
+--
+-- @since 0.0.1.0
+instance Display Nonce where
+  displayBuilder = Builder.fromText . nonceToHexText
+
+-- | ⚠️  Be prudent with what you do with it!
+--
+-- @since 0.0.1.0
+instance Show Nonce where
+  show = BS.unpackChars . nonceToHexByteString
+
+-- | Generate a new random nonce.
+-- Only use it once per exchanged message.
+--
+-- Do not use this outside of hash creation!
+newNonce :: IO Nonce
+newNonce = do
+  (fPtr :: ForeignPtr CUChar) <- Foreign.mallocForeignPtrBytes (fromIntegral cryptoSecretboxNonceBytes)
+  Foreign.withForeignPtr fPtr $ \ptr ->
+    randombytesBuf (Foreign.castPtr @CUChar @Word8 ptr) cryptoSecretboxNonceBytes
+  pure $ Nonce fPtr
+
+-- | Create a 'Nonce' from a 'StrictByteString' that you have obtained on your own,
+-- usually from the network or disk.
+--
+-- @since 0.0.1.0
+nonceFromByteString :: StrictByteString -> Nonce
+nonceFromByteString bytestring = unsafeDupablePerformIO $
+  BS.unsafeUseAsCStringLen bytestring $ \(outsideNoncePtr, _) -> do
+    nonceForeignPtr <- BS.mallocByteString @CChar (fromIntegral cryptoSecretboxNonceBytes)
+    Foreign.withForeignPtr nonceForeignPtr $ \noncePtr ->
+      Foreign.copyArray outsideNoncePtr noncePtr (fromIntegral cryptoSecretboxNonceBytes)
+    pure $ Nonce (Foreign.castForeignPtr @CChar @CUChar nonceForeignPtr)
```

Comment:
Will _any_ `StrictByteString` do? I assume there's a size requirement.

For the point of view of 'reading from network or disk', I'd rather have fewer footguns by doing something like:

```
nonceFromHandle :: Handle StrictByteString -> (Maybe Nonce -> IO r) -> IO r
```

with a giant 'do not leak this nonce no we mean it'.